### PR TITLE
AK+Everywhere: Add and use an exact double parsing method

### LINKED ISSUE #14691: LibJS: Stop using strtod() since it's not safe on platforms that honor the system locale
The number `1.5` will parse as `1` in LibJS on systems where the current locale uses `,` as a decimal separator.

I noticed this when running Ladybird on Linux, where I had `LC_NUMERIC=sv_SE.UTF-8` set.

LibJS uses `strtod()` in a number of places, and we'll want to switch them all to do locale-agnostic parsing.

### PR DESCRIPTION
**Pull request by davidot**
It is based on https://github.com/fastfloat/fast_float, or the paper at https://arxiv.org/abs/2101.11408

Oh and I just found this explanation: https://nigeltao.github.io/blog/2020/eisel-lemire.html
and https://nigeltao.github.io/blog/2020/parse-number-f64-simple.html

There is a number of things I need to figure out to make this as nice and fast as possible:
- [x] Ask the fast_float people about a weird edge case I haven't been able to trigger, asked see https://github.com/fastfloat/fast_float/issues/146
- [x] Add support for floats
  ~- [ ] Is this even needed?~
- [x] Add faster number parsing
- [x] Optimize until it is fast enough
- [x] Determine all the spots in serenity where this should be used
- [x] Make a nice and easy to use API
- [x] Determine how we generate the tables needed, because I would like to generate them, but it's too slow for constexpr it looks like
~~- [ ] Maybe the entire thing can be constexpr...~~ This would require moving everything to headers so that's for future someone

So if you know about a spot in serenity which does string -> double conversion and is not fixme'd or fixed in this PR leave a comment.

Will fix #14691 once implement in LibJS

Part of the test set It is parsing all the floating points in https://github.com/fastfloat/supplemental_test_files correctly

### CODE REVIEW

**Comment by davidot** [comment]
The changes should be done, just need to do the final benchmarks and rewrite some commit messages

**Comment by davidot** [comment]
Benchmark results:
float-bench

Some remarks:
For easy cases (generated with https://gist.github.com/davidot/22eed5fdd526d027ad4d2a4388b0aef4) 
the new parser is not faster in fact it seems a tiny bit slower.

But on the other tests the new method crushes the current implementations, and these also gave wrong results for over 50% of the dataset from https://github.com/fastfloat/supplemental_test_files/tree/main/data
And the performance is much more predictable.

The current SVG number parser just crashed for integers outside of `int` range. So it couldn't even run the benchmarks, but with that fixed it was roughly the same speed as the json one.

The size difference is visible in libcore, the big part is a 651 array with u128, taking 10461 bytes so 10kb. (just noticed the difference is 0.1 mb I shouldn't have used `-h`)

**Comment by davidot** [comment]
@AtkinsSJ You might know whether I have used proper whitespace trimming and implementation for the LibWeb new double parser usage

Also if you happen to know more places where we parse doubles/floats in LibWeb/LibJS/Anything else used by the browser let me know. I checked all the `strtod`, `strtof` and parsers I could find.

File: Userland/Libraries/LibWeb/HTML/HTMLInputElement.cpp [hunk 1]
```diff
@@ -446,9 +446,9 @@ String HTMLInputElement::value_sanitization_algorithm(String value) const
         }
     } else if (type_state() == HTMLInputElement::TypeAttributeState::Number) {
         // If the value of the element is not a valid floating-point number, then set it to the empty string instead.
-        char* end_ptr;
-        auto val = strtod(value.characters(), &end_ptr);
-        if (!isfinite(val) || *end_ptr)
+        // FIXME: The spec does not specify if we need to trim whitespace.
```
Comment (AtkinsSJ) [inline Userland/Libraries/LibWeb/HTML/HTMLInputElement.cpp:449]: I think we can trim the whitespace here and in the HTMLProgressElement. I found this:

https://html.spec.whatwg.org/multipage/common-microsyntaxes.html#rules-for-parsing-floating-point-number-values
Specifically step 6:
> [Skip ASCII whitespace](https://infra.spec.whatwg.org/#skip-ascii-whitespace) within input given position.

Also the **Example** box in https://html.spec.whatwg.org/multipage/form-control-infrastructure.html#concept-fe-value mentions that a UA could trim whitespace on input elements in general if it wanted.

File: Userland/Libraries/LibWeb/HTML/HTMLProgressElement.cpp [hunk 2]
```diff
@@ -47,15 +47,18 @@ void HTMLProgressElement::progress_position_updated()
...
 double HTMLProgressElement::value() const
 {
-    auto value_characters = attribute(HTML::AttributeNames::value).characters();
+    auto const& value_characters = attribute(HTML::AttributeNames::value);
     if (value_characters == nullptr)
         return 0;
 
-    auto parsed_value = strtod(value_characters, nullptr);
-    if (!isfinite(parsed_value) || parsed_value < 0)
+    // FIXME: Do we need to trim whitespace here?
```
Comment (AtkinsSJ) [inline Userland/Libraries/LibWeb/HTML/HTMLProgressElement.cpp:54]: Yes! See other comment.

File: Userland/Libraries/LibWeb/HTML/HTMLProgressElement.cpp [hunk 3]
```diff
@@ -69,15 +72,18 @@ void HTMLProgressElement::set_value(double value)
...
 double HTMLProgressElement::max() const
 {
-    auto max_characters = attribute(HTML::AttributeNames::max).characters();
+    auto const& max_characters = attribute(HTML::AttributeNames::max);
     if (max_characters == nullptr)
         return 1;
 
-    auto parsed_value = strtod(max_characters, nullptr);
-    if (!isfinite(parsed_value) || parsed_value <= 0)
+    // FIXME: Do we need to trim whitespace here?
```
Comment (AtkinsSJ) [inline Userland/Libraries/LibWeb/HTML/HTMLProgressElement.cpp:79]: Yes! See other comment.

**Review by AtkinsSJ — CHANGES_REQUESTED** [review]
I'm still a bit iffy on replacing individual spec-based implementations with a single one. But at the same time, the CSS float format at least seems to be the "standard" one so it won't actually break anything. I expect the others are too. And that speed up is nice!

**Comment by davidot** [comment]
> I'm still a bit iffy on replacing individual spec-based implementations with a single one.

I get that, are you more worried about the supported format or the actual conversion to the appropriate value?
I intentionally made the supported format of the new one as wide as possible so it should except whatever you throw at it. 
And to make sure we actually get the correct value we would have to have 3+ implementations of something like this algorithm.
Even though I'm sure things like `1.239487e97` are barely used in CSS, giving the incorrect value feels like it could cause very hard to spot problems.

I initially wanted to leave the spec comments with a comment saying this amounts to the same format, but that felt like it would just be a trust me comment.

**Comment by AtkinsSJ** [comment]
> > I'm still a bit iffy on replacing individual spec-based implementations with a single one.
> 
> I get that, are you more worried about the supported format or the actual conversion to the appropriate value? I intentionally made the supported format of the new one as wide as possible so it should except whatever you throw at it. And to make sure we actually get the correct value we would have to have 3+ implementations of something like this algorithm. Even though I'm sure things like `1.239487e97` are barely used in CSS, giving the incorrect value feels like it could cause very hard to spot problems.
> 
> I initially wanted to leave the spec comments with a comment saying this amounts to the same format, but that felt like it would just be a trust me comment.

Being too forgiving with formats can be a problem in itself. But, I think we're good here at least for CSS. It always examines the input first to make sure it's a number before consuming one, and treats the float parsing as infallible. So it should be impossible for it to try and convert an invalid-for-css string to a float.

I think it's all fine. I just find specs reassuring! :sweat_smile:

File: Userland/Libraries/LibWeb/HTML/HTMLInputElement.cpp (hunk 1 above)
Comment (davidot) [reply Userland/Libraries/LibWeb/HTML/HTMLInputElement.cpp:449]: Nice catch, found that spec but only looked for if the format was compatible, whitespace doesn't count in format right :facepalm:

I made them all trim the whitespace and added a link to the spec.

File: Userland/Libraries/LibC/stdlib.cpp [hunk 4]
```diff
@@ -182,6 +184,141 @@ inline int generate_unique_filename(char* pattern, size_t suffix_length, Callbac
...
+    // First, they decompose the input string into three parts:
+    char* parse_ptr = const_cast<char*>(str);
+
+    // An initial, possibly empty, sequence of white-space characters (as specified by isspace())
+    strtons(parse_ptr, &parse_ptr);
+
+    // FIXME: This strlen can be very expensive if we are parsing the first value of a long string
+    //        we can also stop at the first white space or other characters which will never occur
+    //        in a floating point.
+    auto length = strlen(parse_ptr);
```
Comment (davidot) [inline Userland/Libraries/LibC/stdlib.cpp:243]: This `strlen` can certainly cause a performance regression for any usage of strtod on a huge string. However I looked through the usages in the repo and those all look fine. 
However ports or future programs might use this so if anyone is super concerned we can add a special version which checks for `\0` instead of `end`

File: Userland/Libraries/LibGfx/Color.cpp [hunk 5]
```diff
@@ -63,7 +64,12 @@ static Optional<Color> parse_rgba_color(StringView string)
     auto g = parts[1].to_int().value_or(256);
     auto b = parts[2].to_int().value_or(256);
 
-    double alpha = strtod(parts[3].to_string().characters(), nullptr);
+    double alpha = 0;
+    char const* start = parts[3].characters_without_null_termination();
+    auto alpha_result = parse_first_floating_point(start, start + parts[3].length());
```
Comment (ADKaster) [inline Userland/Libraries/LibGfx/Color.cpp:69]: wait, is it intentional that `Color::from_string("rgba(banana, apple, pomegranate, pineapple)"sv)` is a value of `Color(256,256,256,0)`?

File: Userland/Libraries/LibJS/Runtime/GlobalObject.cpp [hunk 6]
```diff
@@ -226,10 +227,20 @@ JS_DEFINE_NATIVE_FUNCTION(GlobalObject::parse_float)
...
     auto input_string = TRY(vm.argument(0).to_string(vm));
     auto trimmed_string = MUST(trim_string(vm, js_string(vm, input_string), TrimMode::Left));
-    for (size_t length = trimmed_string.length(); length > 0; --length) {
-        auto number = MUST(Value(js_string(vm, trimmed_string.substring(0, length))).to_number(vm));
-        if (!number.is_nan())
-            return number;
+    auto result = parse_first_floating_point<double>(trimmed_string.characters(), trimmed_string.characters() + trimmed_string.length());
+    if (result.parsed_value())
+        return result.value;
+    if (trimmed_string.is_empty())
```
Comment (ADKaster) [inline Userland/Libraries/LibJS/Runtime/GlobalObject.cpp:233]: if the trimmed string is empty, wouldn't we want to check that before trying to parse it as a float? :thinking:

Comment (davidot) [reply Userland/Libraries/LibJS/Runtime/GlobalObject.cpp]: Yeah I guess that's better, I think I put the empty check here to be able to safely use `trimmed_string[0]` below

File: Userland/Libraries/LibGfx/Color.cpp (hunk 5 above)
Comment (davidot) [reply Userland/Libraries/LibGfx/Color.cpp:69]: I don't know, but that's what the code does already :shrug:

Comment (davidot) [reply Userland/Libraries/LibGfx/Color.cpp:69]: The actual CSS parser has its own implementation of rgba which doesn't have this behaviour. However Color::from_string which can call this method is used in a bunch of other spots unguarded so this is probably incorrect.

File: Userland/Libraries/LibC/stdlib.cpp (hunk 4 above)
Comment (davidot) [reply Userland/Libraries/LibC/stdlib.cpp]: The `strlen` is gone and performance of `strtod` and `strtof` is like it was before for the most common cases, and sped up for long cases (defenitely on long strings with the a small number at the beginning)

**Comment by davidot** [comment]
Screenshot from 2022-10-22 02-34-44

These are the new `strtod` and `strtof` performance numbers, so no regression for common cases is visible anymore and the long cases are even faster!

File: AK/FloatingPointStringConversions.h [hunk 7]
```diff
@@ -0,0 +1,85 @@
...
+    T value {};
+
+    [[nodiscard]] bool parsed_value() const
+    {
+        // All other errors do indicate out of range but did produce a valid value.
+        return error != FloatingPointError::NoOrInvalidInput;
+    }
+};
+
+/// This function finds the first floating point within [start, end). The accepted format is
```
Comment (davidot) [inline AK/FloatingPointStringConversions.h:40]: Just looking at the code doc and method names is already appreciated!

File: Userland/Libraries/LibJS/Tests/test-common.js [hunk 8]
```diff
@@ -306,14 +306,30 @@ class ExpectationError extends Error {
                     threw = false;
                 } catch (e) {
                     if (typeof value === "string") {
-                        this.__expect(e.message.includes(value));
+                        this.__expect(
+                            e.message.includes(value),
+                            `Expected ${this.target.toString()} to throw and message to include ${value} but message ${
+                                e.message
+                            } did not contain it`
```
Comment (linusg) [inline Userland/Libraries/LibJS/Tests/test-common.js:313]: Should probably quote `value` and `e.message` here to make it clear these are strings

File: AK/String.cpp [hunk 9]
```diff
@@ -182,6 +182,17 @@ template Optional<u32> String::to_uint(TrimWhitespace) const;
...
 
+#ifndef KERNEL
+template<typename T>
+Optional<T> String::to_double(TrimWhitespace trim_whitespace) const
+{
+    return StringUtils::convert_to_double<T>(*this, trim_whitespace);
+}
+
+template Optional<double> String::to_double(TrimWhitespace) const;
+template Optional<float> String::to_double(TrimWhitespace) const;
```
Comment (linusg) [inline AK/String.cpp:193]: Shouldn't this be called `to_float()`? Or something like `to_floading_point()`, `too_double<float>()` seems kinda weird :sweat_smile:

File: AK/StringUtils.cpp [hunk 10]
```diff
@@ -232,6 +233,23 @@ template Optional<u16> convert_to_uint_from_octal(StringView str, TrimWhitespace
 template Optional<u32> convert_to_uint_from_octal(StringView str, TrimWhitespace);
 template Optional<u64> convert_to_uint_from_octal(StringView str, TrimWhitespace);
 
+#ifndef KERNEL
+template<typename T>
+Optional<T> convert_to_double(StringView str, TrimWhitespace trim_whitespace)
```
Comment (linusg) [inline AK/StringUtils.cpp:238]: Likewise, wouldn't put "double" in the name here given it's templated.

File: AK/StringView.cpp [hunk 11]
```diff
@@ -236,6 +236,15 @@ template Optional<long> StringView::to_uint() const;
...
 
 #ifndef KERNEL
+template<typename T>
+Optional<T> StringView::to_double(TrimWhitespace trim_whitespace) const
+{
+    return StringUtils::convert_to_double<T>(*this, trim_whitespace);
+}
+
+template Optional<double> StringView::to_double(TrimWhitespace) const;
+template Optional<float> StringView::to_double(TrimWhitespace) const;
```
Comment (linusg) [inline AK/StringView.cpp:246]: Same here

File: Userland/Libraries/LibC/stdlib.cpp [hunk 12]
```diff
@@ -4,7 +4,9 @@
  * SPDX-License-Identifier: BSD-2-Clause
  */
 
+#include "AK/FloatingPointStringConversions.h"
```
Comment (linusg) [inline Userland/Libraries/LibC/stdlib.cpp:7]: `<>`

File: Userland/Libraries/LibC/stdlib.cpp [hunk 13]
```diff
@@ -182,6 +184,166 @@ inline int generate_unique_filename(char* pattern, size_t suffix_length, Callbac
...
+    if (is_either(parse_ptr, 0, 'i', 'I')) {
+        if (is_either(parse_ptr, 1, 'n', 'N')) {
+            if (is_either(parse_ptr, 2, 'f', 'F')) {
+                parse_ptr += 3;
+                if (is_either(parse_ptr, 0, 'i', 'I')) {
+                    if (is_either(parse_ptr, 1, 'n', 'N')) {
+                        if (is_either(parse_ptr, 2, 'i', 'I')) {
+                            if (is_either(parse_ptr, 3, 't', 'T')) {
+                                if (is_either(parse_ptr, 4, 'y', 'Y')) {
+                                    parse_ptr += 5;
```
Comment (linusg) [inline Userland/Libraries/LibC/stdlib.cpp:198]: :yakkiss:

File: Userland/Libraries/LibC/stdlib.cpp [hunk 14]
```diff
@@ -182,6 +184,166 @@ inline int generate_unique_filename(char* pattern, size_t suffix_length, Callbac
...
+                return true;
+            }
+        }
+    }
+
+    return false;
+}
+
+template<FloatingPoint T>
+static T strtofloatingpoint(char const* str, char** endptr)
```
Comment (linusg) [inline Userland/Libraries/LibC/stdlib.cpp:232]: Since this is not an official LibC function, we can use a name that's actually readable

File: AK/FloatingPointStringConversions.h [hunk 15]
```diff
@@ -0,0 +1,68 @@
...
+/// - An optional sign, both + and - are supported
+/// - 0 or more decimal digits, with leading zeros allowed [1]
+/// - A decimal point '.', which can have no digits after it
+/// - 0 or more decimal digits, unless the first digits [1] doesn't have any digits,
+///   then this must have at least one
+/// - An exponent 'e' or 'E' followed by an optional sign '+' or '-' and at least one digit
+/// This function additionally detects out of range values which have been rounded to
+/// [-]infinity or 0 and gives the next character to read after the floating point.
+template<FloatingPoint T = double>
+FloatingPointParseResults<T> parse_first_floating_point(char const* start, char const* end);
```
Comment (linusg) [inline AK/FloatingPointStringConversions.h:52]: Not sure what the implications would be, but can we change this to take a `StringView`? Seems a bit strange to add a new AK function using c strings in this day and age. I'm sure LibC could wrap its input in one...?

File: Userland/Libraries/LibJS/Runtime/Intl/PluralRules.cpp [hunk 16]
```diff
@@ -21,9 +21,9 @@ PluralRules::PluralRules(Object& prototype)
 ::Locale::PluralOperands get_operands(String const& string)
 {
     // 1.Let n be ! ToNumber(s).
-    char* end { nullptr };
-    auto number = strtod(string.characters(), &end);
-    VERIFY(!*end);
+    auto number_or_error = string.to_double(AK::TrimWhitespace::Yes);
```
Comment (linusg) [inline Userland/Libraries/LibJS/Runtime/Intl/PluralRules.cpp:24]: An `Optional<T>` is not an error - our usual naming for this is `maybe_foo`. But I think you can just skip the assertion and `release_value()`, should be clear from the comment above it

File: Userland/Libraries/LibJS/Runtime/Temporal/AbstractOperations.cpp [hunk 17]
```diff
@@ -421,20 +421,20 @@ ThrowCompletionOr<SecondsStringPrecision> to_seconds_string_precision(VM& vm, Ob
     // 15. If fractionalDigitCount is 1, 2, or 3, then
     if (fractional_digit_count == 1 || fractional_digit_count == 2 || fractional_digit_count == 3) {
         // a. Return the Record { [[Precision]]: fractionalDigitCount, [[Unit]]: "millisecond", [[Increment]]: 10^(3 - fractionalDigitCount) }.
-        return SecondsStringPrecision { .precision = fractional_digit_count, .unit = "millisecond"sv, .increment = (u32)pow(10, 3 - fractional_digit_count) };
+        return SecondsStringPrecision { .precision = fractional_digit_count, .unit = "millisecond"sv, .increment = (u32)pow(10., 3 - fractional_digit_count) };
```
Comment (linusg) [inline Userland/Libraries/LibJS/Runtime/Temporal/AbstractOperations.cpp:424]: Why is this necessary? Won't the compiler cast this the int literal to a double anyway?

File: Userland/Libraries/LibJS/Runtime/AbstractOperations.cpp [hunk 18]
```diff
@@ -1180,21 +1181,22 @@ CanonicalIndex canonical_numeric_index_string(PropertyKey const& property_key, C
...
 
     // Short circuit any string that doesn't start with digits
     if (char first_non_zero = argument.characters()[current_index]; first_non_zero < '0' || first_non_zero > '9')
         return CanonicalIndex(CanonicalIndex::Type::Undefined, 0);
 
     // 2. Let n be ! ToNumber(argument).
-    char* endptr;
-    auto n = Value(strtod(argument.characters(), &endptr));
-    if (endptr != argument.characters() + argument.length())
+    auto double_or_failure = argument.to_double(AK::TrimWhitespace::No);
```
Comment (linusg) [inline Userland/Libraries/LibJS/Runtime/AbstractOperations.cpp:1192]: `maybe_double`

File: Userland/Libraries/LibJS/Runtime/Value.cpp [hunk 19]
```diff
@@ -549,12 +550,11 @@ Optional<Value> string_to_number(StringView string)
         return Value(bigint.to_double());
     }
 
-    char* endptr;
-    auto parsed_double = strtod(text.characters(), &endptr);
-    if (*endptr)
+    auto parsed_double = text.to_double(AK::TrimWhitespace::No);
```
Comment (linusg) [inline Userland/Libraries/LibJS/Runtime/Value.cpp:553]: Oh god, another name for the same thing haha - feel free to ignore the previous comment and call that `parsed_double` as well

File: Userland/Libraries/LibWeb/HTML/HTMLInputElement.cpp [hunk 20]
```diff
@@ -446,9 +446,10 @@ String HTMLInputElement::value_sanitization_algorithm(String value) const
         }
     } else if (type_state() == HTMLInputElement::TypeAttributeState::Number) {
         // If the value of the element is not a valid floating-point number, then set it to the empty string instead.
-        char* end_ptr;
-        auto val = strtod(value.characters(), &end_ptr);
-        if (!isfinite(val) || *end_ptr)
+        // https://html.spec.whatwg.org/multipage/common-microsyntaxes.html#rules-for-parsing-floating-point-number-values
+        // 6. Skip ASCII whitespace within input given position.
+        auto number_or_none = value.to_double(TrimWhitespace::Yes);
```
Comment (linusg) [inline Userland/Libraries/LibWeb/HTML/HTMLInputElement.cpp:451]: not even more names pls

File: Userland/Libraries/LibWeb/SVG/AttributeParser.cpp [hunk 21]
```diff
@@ -379,36 +357,18 @@ float AttributeParser::parse_number()
...
-    if (exponent_sign < 0) {
-        for (int i = 0; i < exponent; ++i) {
-            number /= 10;
-        }
-    } else if (exponent_sign > 0) {
-        for (int i = 0; i < exponent; ++i) {
-            number *= 10;
-        }
-    }
+    auto floating_point_result = parse_first_floating_point<float>(start, start + remaining_source_text.length());
```
Comment (linusg) [inline Userland/Libraries/LibWeb/SVG/AttributeParser.cpp:367]: aaaaaaaaah

File: AK/FloatingPointStringConversions.cpp [hunk 22]
```diff
@@ -0,0 +1,2018 @@
...
+ *
+ * SPDX-License-Identifier: BSD-2-Clause
+ */
+
+#include "FloatingPointStringConversions.h"
+#include "CharacterTypes.h"
+#include "Format.h"
+#include "ScopeGuard.h"
+#include "StringView.h"
+#include "UFixedBigInt.h"
```
Comment (linusg) [inline AK/FloatingPointStringConversions.cpp:12]: `<>`

File: AK/FloatingPointStringConversions.cpp [hunk 23]
```diff
@@ -0,0 +1,2018 @@
...
+
+static constexpr i64 lowest_exponent = -342;
+static constexpr i64 highest_exponent = 308;
+
+constexpr auto pre_compute_table()
+{
+    // Computing this entire table at compile time is slow and hits constexpr
+    // limits, so we just compute a (the simplest) value to make sure the
+    // function is used. This means the code is stand alone in case we need to
+    // regenerate this table.
```
Comment (linusg) [inline AK/FloatingPointStringConversions.cpp:486]: How can it be regenerated?

File: AK/FloatingPointStringConversions.cpp [hunk 24]
```diff
@@ -0,0 +1,2018 @@
...
+
+    if (mantissa == 0 || exponent < FloatingPointRepr::min_power_of_10())
+        return FloatingPointBuilder::zero();
+
+    // Max double value which isn't negative is xe308
+    if (exponent > FloatingPointRepr::max_power_of_10())
+        return FloatingPointBuilder::infinity<T>();
+
+    auto w = mantissa;
+    // normalize the decimal significand w by shifting it so that w ∈ [263 , 264 )
```
Comment (linusg) [inline AK/FloatingPointStringConversions.cpp:1306]: `Normalize`, mismatched brackets at the end

File: Tests/AK/TestFloatingPointParsing.cpp [hunk 25]
```diff
@@ -0,0 +1,415 @@
...
+        EXPECT_EQ(static_cast<double>(value), double_parse_function(#value##sv));      \
+        EXPECT_EQ(-static_cast<double>(value), double_parse_function("-" #value##sv)); \
+    } while (false)
+
+#define does_parse_float_like_cpp(value)                                                \
+    do {                                                                                \
+        float val = float_parse_function(#value##sv);                                   \
+        EXPECT_EQ(static_cast<float>(value##f), val);                                   \
+        EXPECT_EQ(-static_cast<float>(value##f), float_parse_function("-" #value##sv)); \
+    } while (false)
```
Comment (linusg) [inline Tests/AK/TestFloatingPointParsing.cpp:40]: Macros should use SCREAMING_CASE please

**Review by linusg — CHANGES_REQUESTED** [review]
Great work! Only did a very shallow review of the two big commits, chances are I won't understand much of it anyway.

File: AK/String.cpp (hunk 9 above)
Comment (davidot) [reply AK/String.cpp:193]: Yeah I went back and forth on this, since the other functions are called `to_int` and `to_uint` and this might make it easier to find. But I'll change it since I agree it is confusing

File: Userland/Libraries/LibJS/Runtime/Temporal/AbstractOperations.cpp (hunk 17 above)
Comment (davidot) [reply Userland/Libraries/LibJS/Runtime/Temporal/AbstractOperations.cpp:424]: Whoops just left from trying some stuff

**Comment by davidot** [comment]
Sorry for the many names :sweat_smile: 
I made `to_double` and `to_float` on strings non templated, hopefully making it more clear.
And then StringUtils does have a templated version in case you want that.

File: AK/FloatingPointStringConversions.cpp (hunk 23 above)
Comment (davidot) [reply AK/FloatingPointStringConversions.cpp]: With the function above, I'll add that to the comment

File: AK/FloatingPointStringConversions.h (hunk 15 above)
Comment (davidot) [reply AK/FloatingPointStringConversions.h:52]: I made these take `char*` because it's the simplest interface.
We can't just use a StringView in libc, since we would have to do `strlen` and that makes `strtod` regress in performace as you can see above.

I think that with the convenience functions on the String types you barely want to call these functions anyway. If you feel this should be changed we can except for the one not taking and `end`

File: AK/String.cpp (hunk 9 above)
Comment (davidot) [reply AK/String.cpp]: Ended up splitting it to `to_double` and `to_float`

**Comment by davidot** [comment]
CI was flaky, just rebased to rerun it

File: AK/FloatingPointStringConversions.cpp [hunk 26]
```diff
@@ -0,0 +1,2018 @@
...
+
+    if (mantissa == 0 || exponent < FloatingPointRepr::min_power_of_10())
+        return FloatingPointBuilder::zero();
+
+    // Max double value which isn't negative is xe308
+    if (exponent > FloatingPointRepr::max_power_of_10())
+        return FloatingPointBuilder::infinity<T>();
+
+    auto w = mantissa;
+    // Normalize the decimal significand w by shifting it so that w ∈ [2^63 , 2^64)
```
Comment (linusg) [inline AK/FloatingPointStringConversions.cpp:1306]:
```suggestion
    // Normalize the decimal significand w by shifting it so that w ∈ [2^63, 2^64]
```

**Review by linusg — COMMENTED** [review]
Looking good, but I think we still got the mismatched bracket in there

File: AK/FloatingPointStringConversions.cpp (hunk 26 above)
Comment (davidot) [reply AK/FloatingPointStringConversions.cpp:1306]: This is not a mismatched bracket, it means the range from 2^63 included to 2^64 excluded

The space before the comma I will fix though

Comment (davidot) [reply AK/FloatingPointStringConversions.cpp:1306]: I guess I can remove the comment as well if you think it's confusing

Comment (linusg) [reply AK/FloatingPointStringConversions.cpp]: TIL! I've always been more of a programming person than a math person, so I intuitively see it as a typo, but it definitely seems to be common range notation.
